katsdpcontroller/product_config.py

```diff
@@ -183,152 +1318,60 @@ def validate(config):
+        valid_types = STREAM_CLASSES[stream['type']]._valid_src_types
         for i, src in enumerate(src_streams):
-            if src in config['inputs']:
-                src_config = config['inputs'][src]
-            elif src in config['outputs']:
-                src_config = config['outputs'][src]
+            if src in inputs:
+                src_config = inputs[src]
+            elif src in outputs:
+                src_config = outputs[src]
             else:
-                raise ValueError('Unknown source {} in {}'.format(src, name))
-            valid_types = src_valid_types[stream['type']]
-            # Special case: valid options depend on position
-            if stream['type'] == 'sdp.spectral_image':
-                valid_types = [valid_types[i]]
-            if src_config['type'] not in valid_types:
-                raise ValueError('Source {} has wrong type for {}'.format(src, name))
-
-    # Can only have one cam.http stream
-    cam_http = [name for (name, stream) in config['inputs'].items()
-                if stream['type'] == 'cam.http']
-    if len(cam_http) > 1:
-        raise ValueError('Cannot have more than one cam.http stream')
-
-    input_endpoints = {}
-    for name, stream in config['inputs'].items():
-        try:
-            if stream['type'] in ['cbf.baseline_correlation_products',
-                                  'cbf.tied_array_channelised_voltage']:
-                n_endpoints = _url_n_endpoints(stream['url'])
-                input_endpoints[name] = n_endpoints
-                src_stream = stream['src_streams'][0]
-                n_chans = config['inputs'][src_stream]['n_chans']
-                n_chans_per_substream = stream['n_chans_per_substream']
-                if n_chans % n_endpoints != 0:
-                    raise ValueError(
-                        'n_chans ({}) not a multiple of endpoints ({})'.format(
-                            n_chans, n_endpoints))
-                n_chans_per_endpoint = n_chans // n_endpoints
-                if n_chans_per_endpoint % n_chans_per_substream != 0:
-                    raise ValueError(
-                        'channels per endpoints ({}) not a multiple of n_chans_per_substream ({})'
-                        .format(n_chans_per_endpoint, n_chans_per_substream))
-        except ValueError as error:
-            raise ValueError('{}: {}'.format(name, error)) from error
+                raise ValueError(f'Unknown source {src} in {name}')
+            if isinstance(valid_types, collections.abc.Set):
+                valid_type = src_config['type'] in valid_types
+            else:
+                valid_type = src_config['type'] == valid_types[i]
+            if not valid_type:
+                raise ValueError(f'Source {src} has wrong type for {name}')
+
+    have_cam_http = False
+    for name, stream in inputs.items():
+        # It's not possible to convert 2.x simulations to 3.0 because we don't
```

Comment:
It would offend me :-) It would also make this function more complicated, for little benefit since it's easy enough to upgrade kattelmod to produce v3 config files.